cluster/juju/layers/kubernetes-master/actions/create-rbd-pv

```diff
@@ -0,0 +1,211 @@
+    context['RBD_FS'] = action_get_or_default('filesystem').strip()
+    context['PV_MODE'] = action_get_or_default('mode').strip()
+
+    # Ensure our paramters match
+    param_validation = validate_parameters(context['RBD_NAME'],
+                                           context['RBD_FS'],
+                                           context['PV_MODE'])
+    if not param_validation == 0:
+        return
+
+    if not validate_unique_volume_name(context['RBD_NAME']):
+        action_fail('Volume name collision detected. Volume creation Aborted.')
+        return
+
+    context['monitors'] = get_monitors()
+
+    # Invoke creation and format the mount device
+    create_rbd_volume(context['RBD_NAME'],
+                      context['RBD_SIZE'],
+                      context['RBD_FS'])
+
+    # Create a temporary workspace to render our persistentVolume template, and
+    # enlist the RDB based PV we've just created
+    with TemporaryDirectory() as active_working_path:
+        temp_template = '{}/pv.yaml'.format(active_working_path)
+        render('rbd-persistent-volume.yaml', temp_template, context)
+
+        check_call(['kubectl', 'create', '-f', temp_template])
+
+
+def action_get_or_default(key):
+    ''' Convenience method to manage defaults since actions dont appear to
+    properly support defaults '''
+
+    value = action_get(key)
+    if value:
+        return value
+    elif key == 'filesystem':
+        return 'xfs'
+    elif key == 'size':
+        return 0
+    elif key == 'mode':
+        return "ReadWriteOnce"
+    else:
+        return ''
+
+
+def create_rbd_volume(name, size, filesystem):
+    ''' Create the RBD volume in Ceph. Then mount it locally to format it for
+    the requested filesystem.
+
+    :param name - The name of the RBD volume
+    :param size - The size in MB of the volume
+    :param filesystem - The type of filesystem to format the block device
+    '''
+
+    # Create the rbd volume
+    # $ rbd create foo --size 50 --image-feature layering
+    command = ['rbd', 'create', '--size', '{}'.format(size), '--image-feature',
+               'layering', name]
+    check_call(command)
+
+    # Lift the validation sequence to determine if we actually created the
+    # rbd volume
+    if validate_unique_volume_name(name):
+        # we failed to create the RBD volume. whoops
+        action_fail('RBD Volume not listed after creation.')
+        print('We failed to validate the volume we created!!!')
+        # TODO: Actually stop execution here and abort.
+
+    mount = ['rbd', 'map', name]
+    device_path = check_output(mount).strip()
+
+    try:
+        format_command = ['mkfs.{}'.format(filesystem), device_path]
+        check_call(format_command)
+        unmount = ['rbd', 'unmap', name]
+        check_call(unmount)
+    except CalledProcessError:
+        action_fail('Failed to format filesystem and unmount.'
+                    ' RDB created but not enlisted.')
+
+
+def get_monitors():
```

Comment:
would it be useful to have the path as a default parameter? Do we ever read different ceph.conf files?